File: benchmarks/knucleotide.go
package main

import "fmt"

func main() {
	n := 200000
	k := 10
	cs := make([]byte, n)
	seed := int64(42)
	for i := 0; i < n; i++ {
		seed = (seed*1103515245 + 12345) % 2147483648
		cs[i] = "ACGT"[(seed/65536)%4]
	}
	counts := make(map[string]int)
	for j := 0; j+k <= n; j++ {
		counts[string(cs[j:j+k])]++
	}
	fmt.Println(len(counts))
}

File: benchmarks/knucleotide.rb
t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)

n = 200000
k = 10
cs = String.new(capacity: n)
seed = 42
chars = "ACGT"

i = 0
while i < n
  seed = (seed * 1103515245 + 12345) % 2147483648
  cs << chars[(seed / 65536) % 4]
  i += 1
end

counts = Hash.new(0)
j = 0
limit = n - k
while j <= limit
  counts[cs[j, k]] += 1
  j += 1
end

puts counts.size
puts "bench_ns=#{Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - t0}"
